File: backend/src/main/java/br/com/calciolari/datahub/imports/infrastructure/storage/LocalRawFileStorage.java
package br.com.calciolari.datahub.imports.infrastructure.storage;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.UncheckedIOException;
import java.nio.file.AtomicMoveNotSupportedException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardCopyOption;
import java.nio.file.StandardOpenOption;
import java.security.DigestInputStream;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HexFormat;
import java.util.Locale;
import java.util.Objects;
import java.util.UUID;

import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

/**
 * Filesystem raw store. Keys look like {@code ab/cd/<sha256>}; never derived from filenames.
 * {@link #putIfAbsent} is idempotent and never overwrites divergent bytes.
 */
@Component
public class LocalRawFileStorage implements RawFileStorage {

	private final Path root;

	public LocalRawFileStorage(@Value("${datahub.raw-storage.root}") String root) {
		this.root = Path.of(Objects.requireNonNull(root, "root")).toAbsolutePath().normalize();
	}

	public Path root() {
		return root;
	}

	@Override
	public StoredRawFile putIfAbsent(InputStream bytes, RawFileDescriptor descriptor) {
		Objects.requireNonNull(bytes, "bytes");
		Objects.requireNonNull(descriptor, "descriptor");
		ensureRoot();

		String sha = descriptor.sha256().toLowerCase(Locale.ROOT);
		String storageKey = storageKeyFor(sha);
		Path target = resolveKey(storageKey);

		synchronized (lockFor(sha)) {
			if (Files.exists(target)) {
				verifyExisting(target, descriptor);
				return new StoredRawFile(storageKey, sha, descriptor.byteSize(), false);
			}

			try {
				Files.createDirectories(target.getParent());
				Path temp = target.getParent().resolve(
						"." + sha + "." + UUID.randomUUID() + ".tmp");
				long written = 0L;
				MessageDigest digest = sha256Digest();
				try (InputStream in = new DigestInputStream(bytes, digest);
						OutputStream out = Files.newOutputStream(temp,
								StandardOpenOption.CREATE_NEW, StandardOpenOption.WRITE)) {
					byte[] buffer = new byte[8192];
					int read;
					while ((read = in.read(buffer)) >= 0) {
						out.write(buffer, 0, read);
						written += read;
					}
				}
				String actualSha = HexFormat.of().formatHex(digest.digest());
				if (!sha.equalsIgnoreCase(actualSha)) {
					Files.deleteIfExists(temp);
					throw new RawStorageIntegrityException(
							"stream SHA-256 mismatch: expected " + sha + " got " + actualSha);
				}
				if (written != descriptor.byteSize()) {
					Files.deleteIfExists(temp);
					throw new RawStorageIntegrityException(
							"stream size mismatch: expected " + descriptor.byteSize() + " got " + written);
				}

				try {
					Files.move(temp, target, StandardCopyOption.ATOMIC_MOVE);
				}
				catch (AtomicMoveNotSupportedException ex) {
					try {
						Files.move(temp, target);
					}
					catch (java.nio.file.FileAlreadyExistsException race) {
						Files.deleteIfExists(temp);
						verifyExisting(target, descriptor);
						return new StoredRawFile(storageKey, sha, descriptor.byteSize(), false);
					}
				}
				catch (java.nio.file.FileAlreadyExistsException ex) {
					Files.deleteIfExists(temp);
					verifyExisting(target, descriptor);
					return new StoredRawFile(storageKey, sha, descriptor.byteSize(), false);
				}

				return new StoredRawFile(storageKey, sha, descriptor.byteSize(), true);
			}
			catch (IOException ex) {
				throw new UncheckedIOException("failed to store raw artifact " + sha, ex);
			}
		}
	}

	private static Object lockFor(String sha) {
		return sha.intern();
	}

	@Override
	public InputStream openVerified(String storageKey, String expectedSha256, long expectedSize) {
		Path path = resolveKey(storageKey);
		if (!Files.isRegularFile(path)) {
			throw new RawStorageIntegrityException("missing raw artifact: " + storageKey);
		}
		try {
			long size = Files.size(path);
			if (size != expectedSize) {
				throw new RawStorageIntegrityException(
						"size mismatch for " + storageKey + ": expected " + expectedSize + " got " + size);
			}
			MessageDigest digest = sha256Digest();
			try (InputStream in = new DigestInputStream(Files.newInputStream(path), digest)) {
				in.transferTo(OutputStream.nullOutputStream());
			}
			String actual = HexFormat.of().formatHex(digest.digest());
			if (!expectedSha256.equalsIgnoreCase(actual)) {
				throw new RawStorageIntegrityException(
						"hash mismatch for " + storageKey + ": expected " + expectedSha256 + " got " + actual);
			}
			return Files.newInputStream(path);
		}
		catch (IOException ex) {
			throw new UncheckedIOException("failed to open raw artifact " + storageKey, ex);
		}
	}

	@Override
	public boolean exists(String storageKey) {
		return Files.isRegularFile(resolveKey(storageKey));
	}

	public static String storageKeyFor(String sha256) {
		String sha = sha256.toLowerCase(Locale.ROOT);
		if (sha.length() != 64) {
			throw new IllegalArgumentException("sha256 must be 64 hex chars");
		}
		return sha.substring(0, 2) + "/" + sha.substring(2, 4) + "/" + sha;
	}

	private void verifyExisting(Path target, RawFileDescriptor descriptor) {
		try {
			long size = Files.size(target);
			if (size != descriptor.byteSize()) {
				throw new RawStorageIntegrityException(
						"existing artifact size mismatch for " + descriptor.sha256()
								+ ": expected " + descriptor.byteSize() + " got " + size);
			}
			MessageDigest digest = sha256Digest();
			try (InputStream in = new DigestInputStream(Files.newInputStream(target), digest)) {
				in.transferTo(OutputStream.nullOutputStream());
			}
			String actual = HexFormat.of().formatHex(digest.digest());
			if (!descriptor.sha256().equalsIgnoreCase(actual)) {
				throw new RawStorageIntegrityException(
						"existing artifact hash mismatch for key of " + descriptor.sha256()
								+ ": got " + actual);
			}
		}
		catch (IOException ex) {
			throw new UncheckedIOException(ex);
		}
	}

	private Path resolveKey(String storageKey) {
		Path resolved = root.resolve(storageKey).normalize();
		if (!resolved.startsWith(root)) {
			throw new SecurityException("storage key escapes root: " + storageKey);
		}
		return resolved;
	}

	private void ensureRoot() {
		try {
			Files.createDirectories(root);
		}
		catch (IOException ex) {
			throw new UncheckedIOException(ex);
		}
	}

	private static MessageDigest sha256Digest() {
		try {
			return MessageDigest.getInstance("SHA-256");
		}
		catch (NoSuchAlgorithmException ex) {
			throw new IllegalStateException(ex);
		}
	}
}

File: backend/src/Calciolari.DataHub/Imports/Infrastructure/Storage/LocalRawFileStorage.cs
using System.Collections.Concurrent;
using System.Security.Cryptography;

namespace Calciolari.DataHub.Imports.Infrastructure.Storage;

/// <summary>
/// Filesystem raw store. Keys look like ab/cd/sha256; never derived from filenames.
/// PutIfAbsent is idempotent and never overwrites divergent bytes.
/// </summary>
public sealed class LocalRawFileStorage : IRawFileStorage
{
    private readonly string _root;
    private static readonly ConcurrentDictionary<string, object> Locks = new(StringComparer.Ordinal);

    public LocalRawFileStorage(string root)
    {
        ArgumentNullException.ThrowIfNull(root);
        _root = Path.GetFullPath(root);
    }

    public string Root => _root;

    public StoredRawFile PutIfAbsent(Stream bytes, RawFileDescriptor descriptor)
    {
        ArgumentNullException.ThrowIfNull(bytes);
        ArgumentNullException.ThrowIfNull(descriptor);
        EnsureRoot();

        var sha = descriptor.Sha256.ToLowerInvariant();
        var storageKey = StorageKeyFor(sha);
        var target = ResolveKey(storageKey);

        lock (LockFor(sha))
        {
            if (File.Exists(target))
            {
                VerifyExisting(target, descriptor);
                return new StoredRawFile(storageKey, sha, descriptor.ByteSize, false);
            }

            try
            {
                Directory.CreateDirectory(Path.GetDirectoryName(target)!);
                var temp = Path.Combine(Path.GetDirectoryName(target)!, "." + sha + "." + Guid.NewGuid() + ".tmp");
                long written = 0;
                using (var digest = IncrementalHash.CreateHash(HashAlgorithmName.SHA256))
                using (var output = new FileStream(temp, FileMode.CreateNew, FileAccess.Write, FileShare.None))
                {
                    var buffer = new byte[8192];
                    int read;
                    while ((read = bytes.Read(buffer, 0, buffer.Length)) > 0)
                    {
                        output.Write(buffer, 0, read);
                        digest.AppendData(buffer.AsSpan(0, read));
                        written += read;
                    }

                    var actualSha = Convert.ToHexString(digest.GetHashAndReset()).ToLowerInvariant();
                    if (!sha.Equals(actualSha, StringComparison.OrdinalIgnoreCase))
                    {
                        File.Delete(temp);
                        throw new RawStorageIntegrityException(
                            "stream SHA-256 mismatch: expected " + sha + " got " + actualSha);
                    }
                }

                if (written != descriptor.ByteSize)
                {
                    File.Delete(temp);
                    throw new RawStorageIntegrityException(
                        "stream size mismatch: expected " + descriptor.ByteSize + " got " + written);
                }

                try
                {
                    File.Move(temp, target);
                }
                catch (IOException)
                {
                    if (File.Exists(temp))
                    {
                        File.Delete(temp);
                    }

                    if (File.Exists(target))
                    {
                        VerifyExisting(target, descriptor);
                        return new StoredRawFile(storageKey, sha, descriptor.ByteSize, false);
                    }

                    throw;
                }

                return new StoredRawFile(storageKey, sha, descriptor.ByteSize, true);
            }
            catch (IOException ex)
            {
                throw new IOException("failed to store raw artifact " + sha, ex);
            }
        }
    }

    public Stream OpenVerified(string storageKey, string expectedSha256, long expectedSize)
    {
        var path = ResolveKey(storageKey);
        if (!File.Exists(path))
        {
            throw new RawStorageIntegrityException("missing raw artifact: " + storageKey);
        }

        try
        {
            var size = new FileInfo(path).Length;
            if (size != expectedSize)
            {
                throw new RawStorageIntegrityException(
                    "size mismatch for " + storageKey + ": expected " + expectedSize + " got " + size);
            }

            var actual = Sha256File(path);
            if (!expectedSha256.Equals(actual, StringComparison.OrdinalIgnoreCase))
            {
                throw new RawStorageIntegrityException(
                    "hash mismatch for " + storageKey + ": expected " + expectedSha256 + " got " + actual);
            }

            return File.OpenRead(path);
        }
        catch (IOException ex)
        {
            throw new IOException("failed to open raw artifact " + storageKey, ex);
        }
    }

    public bool Exists(string storageKey) => File.Exists(ResolveKey(storageKey));

    public int WipeAll()
    {
        if (!Directory.Exists(_root))
        {
            return 0;
        }

        var deleted = 0;
        foreach (var entry in Directory.EnumerateFileSystemEntries(_root))
        {
            if (Directory.Exists(entry))
            {
                deleted += Directory.GetFiles(entry, "*", SearchOption.AllDirectories).Length;
                Directory.Delete(entry, true);
            }
            else
            {
                File.Delete(entry);
                deleted++;
            }
        }

        return deleted;
    }

    public static string StorageKeyFor(string sha256)
    {
        var sha = sha256.ToLowerInvariant();
        if (sha.Length != 64)
        {
            throw new ArgumentException("sha256 must be 64 hex chars", nameof(sha256));
        }

        return sha[..2] + "/" + sha[2..4] + "/" + sha;
    }

    private static object LockFor(string sha) => Locks.GetOrAdd(sha, static _ => new object());

    private static void VerifyExisting(string target, RawFileDescriptor descriptor)
    {
        try
        {
            var size = new FileInfo(target).Length;
            if (size != descriptor.ByteSize)
            {
                throw new RawStorageIntegrityException(
                    "existing artifact size mismatch for " + descriptor.Sha256
                    + ": expected " + descriptor.ByteSize + " got " + size);
            }

            var actual = Sha256File(target);
            if (!descriptor.Sha256.Equals(actual, StringComparison.OrdinalIgnoreCase))
            {
                throw new RawStorageIntegrityException(
                    "existing artifact hash mismatch for key of " + descriptor.Sha256
                    + ": got " + actual);
            }
        }
        catch (IOException ex)
        {
            throw new IOException(ex.Message, ex);
        }
    }

    private string ResolveKey(string storageKey)
    {
        var resolved = Path.GetFullPath(Path.Combine(_root, storageKey));
        var rootWithSep = _root.TrimEnd(Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar)
                          + Path.DirectorySeparatorChar;
        if (!resolved.Equals(_root, StringComparison.Ordinal) &&
            !resolved.StartsWith(rootWithSep, StringComparison.Ordinal))
        {
            throw new UnauthorizedAccessException("storage key escapes root: " + storageKey);
        }

        return resolved;
    }

    private void EnsureRoot() => Directory.CreateDirectory(_root);

    private static string Sha256File(string path)
    {
        using var stream = File.OpenRead(path);
        return Convert.ToHexString(SHA256.HashData(stream)).ToLowerInvariant();
    }
}
